Minor adjustments and bug fixing of resampling_in_space

konstntokas opened the pull request:
This PR is related to the previous PR #1098. Further I enhanced the testing of `resample_in_space` to cover the case of a reprojection, where a subsetting in `rectify_dataset` is performed.  

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [ ] ~Add docstrings and API docs for any new/modified user-facing classes and functions~
* [ ] ~New/modified features documented in `docs/source/*`~
* [x] Changes documented in `CHANGES.md`
* [x] GitHub CI passes
* [x] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

TonioF commented on xcube/core/resampling/rectify.py:
```diff
@@ -138,7 +138,18 @@ def rectify_dataset(
             return None
         if source_ds_subset is not source_ds:
-            source_gm = GridMapping.from_dataset(source_ds_subset)
+            # Handles the case where 2D coordinates are already present in the dataset,
+            # which typically occurs after a reprojection performed by the
+            # `resample_in_space` function.
+            if "transformed_x" and "transformed_y" in source_ds_subset:
+                source_gm = GridMapping.from_coords(
+                    source_ds_subset.transformed_x,
+                    source_ds_subset.transformed_y,
+                    crs=source_gm.crs,
+                    tile_size=source_gm.tile_size,
+                )
+            else:
+                source_gm = GridMapping.from_dataset(source_ds_subset)
```
This works, but I have the feeling this is a workaround. As the coordinates are included in source_ds, it should be possible to call 

GridMapping.from_dataset(source_ds_subset, prefer_crs=source_gm.crs) 

and get the correct result.

konstntokas replied:
I changed the setup up the `source_ds` in `resample_in_space`. I took Sentinel-3 data as inspiration (see notebook https://github.com/xcube-dev/xcube/blob/main/examples/notebooks/resampling/rectify_dataset.ipynb).

TonioF reviewed (changes requested):
The new solution would cause problems with data variables that have a gridmapping attribute that points to a crs variable which might not exist anymore after these changes. I still prefer the solution of a variable "transformed_spatial_ref". However, if you prefer your way, please make sure the attributes are not wrong.

konstntokas commented:
> The new solution would cause problems with data variables that have a gridmapping attribute that points to a crs variable which might not exist anymore after these changes. I still prefer the solution of a variable "transformed_spatial_ref". However, if you prefer your way, please make sure the attributes are not wrong.

Thank you that you are looking into it so carefully!

I need to remove the original coordinates, because if the source ds has `x` and `y`, then either `x` and `y` or `transformed_x` and `transformed_y` will be added to the `projected_coords` in the function `xcube.core.gridmapping.cfconv.get_dataset_grid_mapping_proxis`, depending if `x` or `transformed_x` (same for y) comes first in the list `potential_coord_vars`. I therefore prefer to setup a clean dataset in `resample_in_space`, where only one set of coordinates can be found. 

I thus would like to keep my suggestion and also delete all grid_mappings from the data variables in the source_ds. I also added a `grid_mapping` attribute to the unittest.

konstntokas commented:
> > I need to remove the original coordinates, because if the source ds has `x` and `y`, then either `x` and `y` or `transformed_x` and `transformed_y` will be added to the `projected_coords` in the function `xcube.core.gridmapping.cfconv.get_dataset_grid_mapping_proxis`, depending if `x` or `transformed_x` (same for y) comes first in the list `potential_coord_vars`. I therefore prefer to setup a clean dataset in `resample_in_space`, where only one set of coordinates can be found.
> 
> Do you really have to? For testing, I kept existing variables crs and spatial_ref and added a new one transformed_spatial_ref. Now, even though the source_gm retrieved from GridMapping.from_dataset() was the UTM one, the tests eventually succeeded.

I changed it to `transformed_spatial_ref` without deleting the previous crs or spatail_ref variable/coordinate. The the CI fails in the newly added test, if actual subsetting is performed, see second test in `test_spatial.test_reproject_utm()`.

TonioF commented:
You're right, when I executed this, I had left in the line where x and y were removed from the dataset. I now would accept your solution from before the last two commits. It's still not perfect, but I wouldn't be able to come up with a better solution.

konstntokas commented:
@forman can you have a look onto our discussion and my suggested changes. I think a second opinion is useful here. Thank you!

forman commented on CHANGES.md:
```diff
@@ -55,6 +55,9 @@
 
+* The function `xcube.core.resample.resample_in_space()` now supports the parameter
+  `source_ds_subset=True` when calling `rectify_dataset`. This feature enables
+  performing the reprojection exclusively on the congruent subset of the dataset.
```
```suggestion
  performing the reprojection exclusively on the spatially congruent subset of 
  the dataset.
```

forman replied:
Not sure, maybe _congruent_ applies only to the spatial domain. Then ignore my suggestion.

forman commented on xcube/core/resampling/spatial.py:
```diff
@@ -208,11 +208,33 @@ def resample_in_space(
     # transform the source_gm so its CRS matches the target CRS:
     transformed_source_gm = source_gm.transform(target_gm.crs, xy_res=target_gm.xy_res)
+    source_ds = source_ds.drop_vars(source_gm.xy_dim_names)
+    if "crs" in source_ds:
+        source_ds = source_ds.drop_vars("crs")
+    if "spatial_ref" in source_ds:
+        source_ds = source_ds.drop_vars("spatial_ref")
```
This may not be right in cases. You should just drop the grid mapping variable, no matter how it is named. `crs` and `spatial_ref` are common names, but their names are not a sufficient condition to identify the grid mapping variable. There can even be more than one such variables if multiple CRSes are used.

konstntokas replied:
I now go through the `grid_mapping` in the data variable attributes, and  not only delete the attribute with key key `grid_mapping`, but also drop the grid mapping variable. Afterwards I still check for the grid mapping variable named with `crs` and `spatial_ref`. For Sentinel 2 jp2 files e.g. the dataset has the coordinate `spatial_ref`, but the data variable has no attribute `grid_mapping`.

forman replied:
Thats good.

forman commented on xcube/core/gridmapping/cfconv.py:
```diff
@@ -72,7 +72,6 @@ def get_dataset_grid_mapping_proxies(
     grid_mapping_proxies: dict[Union[Hashable, None], GridMappingProxy] = dict()
 
     # Find any grid mapping variables by CF 'grid_mapping' attribute
-    #
```
I like these spacers so the comment is not glued to the next line, but still connected.

konstntokas replied:
No preference here. I changed it back. :)

forman reviewed (approved):
@TonioF @konstntokas 

I approve unless I overlooked the place where you added `transformed_spatial_ref`. I'd be against it. Instead replace the existing source grid mapping from. If you need a history, place a record into the global `history` attribute.
